File: ssi-chat/src/chat_application/auth/retrieval_scope.py
"""Chat retrieval LOB scope derived from the logged-in subject (issue #63)."""

from __future__ import annotations

from typing import Any

from chat_application.auth.capabilities import COMPLIANCE_ROLES
from chat_application.auth.subject import Subject

_LOB_KEYS = ("owning_lob", "lob", "instruction_owning_lob")


def allowed_retrieval_lobs(subject: Subject | None) -> frozenset[str] | None:
    """LOBs the subject may see in graph/direct retrieval.

    Returns:
        ``None`` — unscoped (compliance / platform admin / missing subject)
        empty frozenset — no LOB entitlement
        non-empty frozenset — FO desk lob or MO covering_lobs
    """
    if subject is None:
        return None

    roles = set(subject.roles)
    groups = set(subject.groups)
    if roles & COMPLIANCE_ROLES or "COMPLIANCE" in groups:
        return None

    if "MIDDLE_OFFICE" in groups:
        return frozenset(
            lob.strip().upper()
            for lob in subject.covering_lobs
            if isinstance(lob, str) and lob.strip()
        )

    if subject.lob and str(subject.lob).strip():
        return frozenset({str(subject.lob).strip().upper()})

    return frozenset()


def row_owning_lob(row: dict[str, Any]) -> str | None:
    for key in _LOB_KEYS:
        value = row.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip().upper()
    return None


def filter_rows_by_retrieval_lobs(
    rows: list[dict[str, Any]],
    allowed_lobs: frozenset[str] | None,
) -> list[dict[str, Any]]:
    """Drop detail rows whose owning_lob is outside the subject scope.

    Aggregate rows without an owning_lob column are kept (Cypher should already
    have applied the subject filter for counts).
    """
    if allowed_lobs is None:
        return rows
    kept: list[dict[str, Any]] = []
    for row in rows:
        lob = row_owning_lob(row)
        if lob is None or lob in allowed_lobs:
            kept.append(row)
    return kept


__all__ = [
    "allowed_retrieval_lobs",
    "filter_rows_by_retrieval_lobs",
    "row_owning_lob",
]

File: ssi-chat-j/src/main/java/com/sanjuthomas/policypilot/auth/RetrievalScope.java
package com.sanjuthomas.policypilot.auth;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Subject LOB scope for neo4j_direct / graph retrieval (parity with Python {@code
 * allowed_retrieval_lobs}).
 *
 * <ul>
 *   <li>{@code null} — unscoped (compliance / platform admin)
 *   <li>empty set — no LOB entitlement
 *   <li>non-empty — FO desk lob or MO covering_lobs
 * </ul>
 */
public final class RetrievalScope {

  private static final Set<String> COMPLIANCE_ROLES =
      Set.of("COMPLIANCE_ANALYST", "COMPLIANCE_OFFICER", "PLATFORM_ADMIN");

  private RetrievalScope() {}

  public static Set<String> allowedRetrievalLobs(Subject subject) {
    if (subject == null) {
      return null;
    }
    List<String> roles = subject.roles() == null ? List.of() : subject.roles();
    List<String> groups = subject.groups() == null ? List.of() : subject.groups();
    for (String role : roles) {
      if (role != null && COMPLIANCE_ROLES.contains(role)) {
        return null;
      }
    }
    for (String group : groups) {
      if (group != null && "COMPLIANCE".equalsIgnoreCase(group)) {
        return null;
      }
    }
    for (String group : groups) {
      if (group != null && "MIDDLE_OFFICE".equalsIgnoreCase(group)) {
        Set<String> covering = new LinkedHashSet<>();
        if (subject.coveringLobs() != null) {
          for (String lob : subject.coveringLobs()) {
            if (lob != null && !lob.isBlank()) {
              covering.add(lob.strip().toUpperCase(Locale.ROOT));
            }
          }
        }
        return covering;
      }
    }
    if (subject.lob() != null && !subject.lob().isBlank()) {
      return Set.of(subject.lob().strip().toUpperCase(Locale.ROOT));
    }
    return Set.of();
  }
}
